caffe2/opt/backend_cutting.cc

```diff
@@ -0,0 +1,416 @@
+  if (isa<nom::repr::NeuralNetData>(node_data)) {
+    const auto* nn_tensor = dyn_cast<const nom::repr::NeuralNetData>(node_data);
+    return MakeString("Tensor: ", nn_tensor->getName());
+  } else if (isa<nom::repr::NeuralNetOperator>(node_data)) {
+    const auto* nn_op = dyn_cast<const nom::repr::NeuralNetOperator>(node_data);
+    const auto* op_def = reinterpret_cast<const caffe2::OperatorDef*>(
+        nn_op->getAnnotation()->getSaved());
+    CAFFE_ENFORCE(op_def);
+    return MakeString("Op: ", op_def->type());
+  } else {
+    CAFFE_THROW("Known node");
+  }
+}
+
+void DumpGraph(nom::repr::NNGraph* g) {
+  auto nnprinter = [](typename nom::repr::NNGraph::NodeRef node) {
+    std::map<std::string, std::string> labelMap;
+    assert(node->data() && "Node doesn't have data, can't render it");
+    if (isa<nom::repr::NeuralNetOperator>(node->data())) {
+      auto* op = dyn_cast<nom::repr::NeuralNetOperator>(node->data().get());
+      labelMap["label"] =
+          op->getName() + " (" + std::to_string((unsigned long long)node) + ")";
+      auto* annotation = op->getAnnotation();
+      if (annotation && isa<nom::repr::DeviceAnnotation>(annotation)) {
+        auto device_annotation =
+            dyn_cast<nom::repr::DeviceAnnotation>(annotation);
+        labelMap["label"] += "\\n[" + device_annotation->getDevice() + "]";
+        auto hash = std::hash<std::string>{}(device_annotation->getDevice());
+        std::stringstream hex_stream;
+        hex_stream << std::hex << hash;
+        labelMap["color"] = "#" + hex_stream.str().substr(0, 6);
+        labelMap["fontcolor"] = labelMap["color"];
+      }
+      labelMap["shape"] = "box";
+    } else if (isa<nom::repr::Data>(node->data())) {
+      auto tensor = dyn_cast<nom::repr::NeuralNetData>(node->data().get());
+      labelMap["label"] = tensor->getName();
+      labelMap["label"] += "_" + std::to_string(tensor->getVersion()) + " " +
+          std::to_string((unsigned long long)node);
+    }
+    return labelMap;
+  };
+
+  std::ofstream out("dump.dot");
+  out << nom::converters::convertToDotString(g, nnprinter);
+  out.close();
+}
+
+// Explore the graph in topological order until we hit stopping nodes. This is
+// based on Khan's algorithm:
+// https://en.wikipedia.org/wiki/Topological_sorting#Kahn's_algorithm
+// Precondition: nodes in `current_frontier` must have satisfy `in_degree == 0`
+void Explore(
+    const std::vector<NodeRef>& current_frontier,
+    VisitorContext* context) {
+  std::queue<NodeRef> q;
+  for (const auto n : current_frontier) {
+    q.push(n);
+  }
+
+  while (!q.empty()) {
+    auto node = q.front();
+    q.pop();
+    auto& info = context->infos.at(node);
+
+    // Check if the node is supported, stop exploring further if not supported
+    auto* node_data = node->data().get();
+    if (isa<nom::repr::NeuralNetOperator>(node_data)) {
+      const auto* nn_op =
+          dyn_cast<const nom::repr::NeuralNetOperator>(node_data);
```

Comment:
because this pattern is so common there's actually
`bool nom::repr::nn::is<NeuralNetOperator>(NodeRef* );`
`NeuralNetOperator* nom::repr::nn::get<NeuralNetOperator>(NodeRef* );`

(the `repr::nn` namespace has a couple other useful utilities)